app/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationDao.java

```diff
@@ -0,0 +1,147 @@
+        subpop.setDeleted(existing.isDeleted());
+        
+        mapper.save(subpop);
+        return subpop;
+    }
+
+    @Override
+    public List<Subpopulation> getSubpopulations(StudyIdentifier studyId, boolean allExistingRecordsOnly) {
+        DynamoSubpopulation hashKey = new DynamoSubpopulation();
+        hashKey.setStudyIdentifier(studyId.getIdentifier());
+        
+        DynamoDBQueryExpression<DynamoSubpopulation> query = 
+                new DynamoDBQueryExpression<DynamoSubpopulation>().withHashKeyValues(hashKey);
+        
+        // Get all the records because we only create a default if there are no physical records, 
+        // regardless of the deletion status.
+        List<DynamoSubpopulation> subpops = mapper.query(DynamoSubpopulation.class, query);
+        if (!allExistingRecordsOnly && subpops.isEmpty()) {
+            Subpopulation subpop = createDefaultSubpopulation(studyId);
+            return Lists.newArrayList(subpop);
+        }
+        // Now filter out deleted subpopulations, if requested
+        return subpops.stream()
+            .filter(subpop -> allExistingRecordsOnly || !subpop.isDeleted())
+            .sorted(SPECIFICITY_SORTER).collect(Collectors.toList());
+    }
+    
+    @Override
+    public Subpopulation createDefaultSubpopulation(StudyIdentifier studyId) {
+        DynamoSubpopulation subpop = new DynamoSubpopulation();
+        subpop.setStudyIdentifier(studyId.getIdentifier());
+        subpop.setGuid(studyId.getIdentifier());
+        subpop.setName("Default Consent Group");
+        subpop.setMinAppVersion(0);
+        subpop.setRequired(true);
+        mapper.save(subpop);
+        return subpop;
+    }
+    
+    @Override
+    public Subpopulation getSubpopulation(StudyIdentifier studyId, String guid) {
+        DynamoSubpopulation hashKey = new DynamoSubpopulation();
+        hashKey.setStudyIdentifier(studyId.getIdentifier());
+        hashKey.setGuid(guid);
+        
+        Subpopulation subpop = mapper.load(hashKey);
+        if (subpop == null) {
+            throw new EntityNotFoundException(Subpopulation.class);
+        }
+        return subpop;
+    }
+
+    @Override
+    public Subpopulation getSubpopulationForUser(ScheduleContext context) {
+        List<Subpopulation> found = Lists.newArrayList();
+
+        List<Subpopulation> subpops = getSubpopulations(context.getStudyIdentifier(), false);
+        for (Subpopulation subpop : subpops) {
+            boolean matches = CriteriaUtils.matchCriteria(context, subpop);
+            if (matches) {
+                found.add(subpop);
+            }
+        }
+        if (found.isEmpty()) {
+            throw new EntityNotFoundException(Subpopulation.class);
+        }
+        // Return the most specific match. Eventually we may return them all
+        return found.stream().sorted(SPECIFICITY_SORTER).collect(Collectors.toList()).get(0);
```

Comment:
Would it be better to use Collections.sort() instead? https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#sort-java.util.List-java.util.Comparator-

I find this much more readable than using a stream, and I'm guessing it involves fewer intermediate objects as well.